EyeOnBuzz/Infrastructure/DataSources/GenresDataSource.swift

```diff
@@ -0,0 +1,47 @@
+//
+//  GenresDataSource.swift
+//  EyeOnBuzz
+//
+//  Created by Weverton Couto Timoteo on 1/28/18.
+//  Copyright © 2018 Weverton Couto Timoteo. All rights reserved.
+//
+
+import Foundation
+
+class GenresDataSource {
+    
+    let session = TMDBSession.init()
+    let resultsNodeName = "genres"
+    let dataSourceTarget: DataSourceTarget
+    
+    var list = [Int: String]()
+    
+    init(dataSourceTarget: DataSourceTarget) {
+        self.dataSourceTarget = dataSourceTarget
+    }
+    
+    func fetch() {
+        self.session.get(url: Router.TheMovieDatabaseAPI.genres, parameters: nil, success: { statusCode, headers, response in
+            DispatchQueue.main.async {
+                if let fullResponse = response as? Dictionary<String, Any> {
+                    if let genresResponse = fullResponse[self.resultsNodeName] as? [Dictionary<String, Any>] {
+                        for genreResponse in genresResponse {
+                            let genre = Genre.init(genreResponse)
+                            
+                            self.list[genre.identifier] = genre.name
```

Comment:
Could calling `self` in a closure like this cause any problems?

Reply:
Yes, when you have a closure within outer closure and use `self` we create a strong reference, so it would ever be deallocated. To fix that we should use a `[weak self]` in most outer closure to avoid this strong reference so will be necessary to use `self` with `self?` (optional) as we can't determine if `self` isn't `nil` in "child" closure.